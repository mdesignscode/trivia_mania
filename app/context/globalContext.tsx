"use client";

import storageAvailable from "@/components/localStorageDetection";
import { IGetUserRequest } from "@/models/customRequests";
import User from "@/models/user";
import {
  CATEGORIES,
  DIFFICULTY,
  NEW_PARAMS,
  PROGRESS,
  USERNAME,
  clearQuestionData,
} from "@/utils/localStorage_utils";
import { useUser } from "@clerk/nextjs";
import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { usePathname } from "next/navigation";
import {
  Dispatch,
  SetStateAction,
  createContext,
  useEffect,
  useState,
} from "react";

export interface IUserStatus {
  user: User | null;
  isOnline: boolean;
  isLoaded: boolean;
}

type TFilters = {
  difficulty: string;
  categories: string;
};

export interface IGlobalContext {
  triviaUser: User | null;
  storageIsAvailable: boolean;
  setPlayFilters: (
    cb: (state: { difficulty: string; categories: string }) => {
      difficulty: string;
      categories: string;
    }
  ) => void;
  playFilters: TFilters;
  playUrl: string;
  setPageReady: Dispatch<SetStateAction<boolean>>;
  pageReady: boolean;
  difficultyChoice: Record<string, boolean>;
  setDifficultyChoice: Dispatch<SetStateAction<Record<string, boolean>>>;
  categoryChoice: boolean[];
  setCategoryChoice: Dispatch<SetStateAction<boolean[]>>;
  playerMode: "Guest" | "Signed In";
  setPlayerMode: Dispatch<SetStateAction<"Guest" | "Signed In">>;
  newFilters: boolean;
  setNewFilters: Dispatch<SetStateAction<boolean>>;
}

export const initialGlobalContext: IGlobalContext = {
  triviaUser: null,
  storageIsAvailable: false,
  playFilters: {
    difficulty: "",
    categories: "",
  },
  setPlayFilters: () => {},
  playUrl: "/game?difficulty=&categories=",
  pageReady: false,
  setPageReady: () => {},
  difficultyChoice: {},
  setDifficultyChoice: () => {},
  categoryChoice: [],
  setCategoryChoice: () => {},
  playerMode: "Guest",
  setPlayerMode: () => {},
  setNewFilters: () => {},
  newFilters: false
};

export const GlobalContext =
  createContext<IGlobalContext>(initialGlobalContext);

export function GlobalProvider({ children }: { children: React.ReactNode }) {
  // global state
  // app uses localStorage
  const [storageIsAvailable, setStorageIsAvailable] = useState(false);
  const [triviaUser, setTriviaUser] = useState<User | null>(null);

  const [difficultyChoice, setDifficultyChoice] = useState<
    Record<string, boolean>
  >({
    easy: false,
    medium: false,
    hard: false,
    "all difficulties": false,
  });
  const [categoryChoice, setCategoryChoice] = useState<boolean[]>([]);

  // play navigation link will have current play filters
  const [playFilters, setPlayFilters] = useState({
    difficulty: "",
    categories: "",
  });
  const playUrl = encodeURI(
    `/game?difficulty=${playFilters.difficulty}&categories=${playFilters.categories}`
  );

  const [newFilters, setNewFilters] = useState(false);

  const [pageReady, setPageReady] = useState(false);

  const { user } = useUser();

  const [playerMode, setPlayerMode] = useState<"Guest" | "Signed In">("Guest");

  const path = usePathname();

  useEffect(() => {
    // detect storage feature
    setStorageIsAvailable(storageAvailable());
  }, []);

  const { data, isFetched } = useQuery({
    queryKey: ["getUser"],
    queryFn: async () => {
      try {
        const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
        const url = baseUrl + "users/get";

        const { data } = await axios.post(url, {
          id: user?.id,
        } as IGetUserRequest);

        return data;
      } catch (error) {
        console.log(error);
      }
    },
    initialData: {},
    enabled: !!user,
  });

  useEffect(() => {
    if (/sso-callback/.test(path)) {
      setPageReady(true)
    }
  }, [path])

  // set online user
  useEffect(() => {
    if (isFetched && user) {
      // set online user
      const triviaUser = data;
      setTriviaUser(triviaUser);

      if (storageIsAvailable) {
        // check if current user is previous user
        const prevUserName = localStorage.getItem(USERNAME);

        if (prevUserName) {
          if (prevUserName !== triviaUser.username) {
            // if not previous user, clear localStorage
            clearQuestionData();
            localStorage.setItem(USERNAME, triviaUser.username);
            localStorage.removeItem(PROGRESS);
            localStorage.removeItem(DIFFICULTY);
            localStorage.removeItem(CATEGORIES);
          }
        } else {
          localStorage.setItem(USERNAME, triviaUser.username);
        }
      }
    }
  }, [data, isFetched, storageIsAvailable, user]);

  // get last set filters from local storage and put it in state
  useEffect(() => {
    if (storageIsAvailable) {
      const previousDifficulty = localStorage.getItem(DIFFICULTY);
      const previousCategories = localStorage.getItem(CATEGORIES);

      // first run on client
      if (!previousDifficulty && !previousCategories) {
        localStorage.setItem(NEW_PARAMS, "true");
      }

      if (previousDifficulty) {
        setPlayFilters((state) => ({
          ...state,
          difficulty: previousDifficulty,
        }));

        setDifficultyChoice((state) => ({
          ...state,
          [previousDifficulty]: true,
        }));
      }

      if (previousCategories) {
        setPlayFilters((state) => ({
          ...state,
          categories: previousCategories,
        }));
      }
    }
  }, [storageIsAvailable]);

  // store object
  const store: IGlobalContext = {
    triviaUser,
    storageIsAvailable: storageIsAvailable,
    playFilters,
    setPlayFilters: setPlayFilters,
    playUrl,
    pageReady,
    setPageReady,
    setDifficultyChoice,
    difficultyChoice,
    setCategoryChoice,
    categoryChoice,
    playerMode,
    setPlayerMode,
    newFilters,
    setNewFilters,
  };

  return (
    <GlobalContext.Provider value={store}>{children}</GlobalContext.Provider>
  );
}
