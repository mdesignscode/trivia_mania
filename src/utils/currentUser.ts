import { redirect, type Cookies } from "@sveltejs/kit";
import { validateSessionToken } from "./session";

// check if user is already logged in
export const isSignedIn = async (token: string | undefined, route: string) => {
        // don't redirect auth pages
        const isAuthRoute = ['/login', '/signup'].includes(route);
        if (!token && isAuthRoute) return;

        // user not logged in
        if (!token) throw redirect(302, '/login');

        const { user } = await validateSessionToken(token);
        // user signed in
        if (user) throw redirect(302, '/');

        // no user
        if (!user && isAuthRoute) return;
        else throw redirect(302, '/login');
}

export const getUser = async (cookies: Cookies) => {
        const token = cookies.get('session');
        if (!token) {
                throw redirect(302, '/login');
        }

        const { session, user } = await validateSessionToken(token);
        if (!session || !user) {
                throw redirect(302, '/login');
        }

        const userData = user.dataValues;
        delete userData.password;

        return {
                user: userData,
        };
}

