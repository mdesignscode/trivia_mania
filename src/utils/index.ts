import { browser } from '$app/environment';

export function decodeHTMLEntities(text: string) {
        if (browser) {
                const element = document.createElement("div");
                element.innerHTML = text;
                return element.textContent || "";
        }
}

export const colorMap: { [key: string]: string } = {
        easy: "green",
        medium: "gold",
        hard: "red",
};

export function capitalize(str: string): string {
	if (!str) return '';
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

