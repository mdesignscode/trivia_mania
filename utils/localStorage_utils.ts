export const QUESTIONS_POOL = "questionsPool",
  POOL_INDEX = "poolIndex",
  CURRENT_INDEX = "currentIndex",
  QUESTIONS_LIST = "questionsList",
  QUESTION_ANSWERED = "questionAnswered",
  ANSWERED_QUESTIONS = "answeredQuestions",
  PROGRESS = "progress",
  LAST_ANSWER = "lastAnswer",
  LAST_ANSWER_INDEX = "lastAnswerIndex",
  NEW_PARAMS = "newParams",
  CATEGORIES = "categories",
  DIFFICULTY = "difficulty",
  UNSAVED_DATA = "unsavedData",
  USERNAME = "username"

export function clearQuestionData() {
  localStorage.removeItem(UNSAVED_DATA);
  localStorage.removeItem(QUESTIONS_LIST);
  localStorage.removeItem(QUESTIONS_POOL);
  localStorage.removeItem(POOL_INDEX);
  localStorage.removeItem(CURRENT_INDEX);
  localStorage.removeItem(ANSWERED_QUESTIONS);
  localStorage.removeItem(LAST_ANSWER);
  localStorage.removeItem(LAST_ANSWER_INDEX);
  localStorage.removeItem(ANSWERED_QUESTIONS);
  localStorage.removeItem(NEW_PARAMS);
}
