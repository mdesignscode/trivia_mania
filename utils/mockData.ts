import { CategoryStat, DifficultyStat, IUserStats } from "@/models/interfaces";
import { IQuestion } from "@/models/interfaces";

type TStat = Record<string, number>

export const mockInitialProgress: IUserStats = {
  total: {
    correctAnswered: 1,
    answered: 1,
  },
  easy: {
    answered: 1,
    correctAnswered: 1,
  } as DifficultyStat,
  Science: {
    easy: {
      answered: 1,
      correctAnswered: 1,
    } as DifficultyStat,
  } as CategoryStat,
};

// create mock questions
export const mockQuestions: Record<string, IQuestion> = {
  "63805aab-b589-4256-9493-769eecc23966": {
    "category": "Entertainment: Film",
    "answers": [
      "Batman v Superman: Superapocalypse",
      "Batman v Superman: Black of Knight",
      "Batman v Superman: Dawn of Justice",
      "Batman v Superman: Knightfall"
    ],
    "correctAnswer": "Batman v Superman: Dawn of Justice",
    "id": "63805aab-b589-4256-9493-769eecc23966",
    "question": "Which movie released in 2016 features Superman and Batman fighting?",
    "difficulty": "medium"
  },
  "36240604-80aa-4ddb-bb87-5ce345bfd379": {
    "category": "Entertainment: Film",
    "answers": ["R2-D2", "AA-A", "BB-8", "BB-3"],
    "correctAnswer": "BB-8",
    "id": "36240604-80aa-4ddb-bb87-5ce345bfd379",
    "question": "What is the orange and white bot&#039;s name in &quot;Star Wars: The Force Awakens&quot;?",
    "difficulty": "easy"
  },
  "12c4b40f-7fa9-405d-81de-b64c54a0caea": {
    "category": "Entertainment: Music",
    "answers": [
      "James Blunt - &quot;1973&quot;",
      "Prince - &quot;1999&quot;",
      "The Smashing Pumpkins - &quot;1979&quot;",
      "David Bowie - &quot;1984&quot;"
    ],
    "correctAnswer": "The Smashing Pumpkins - &quot;1979&quot;",
    "id": "12c4b40f-7fa9-405d-81de-b64c54a0caea",
    "question": "Which of these songs was released in 1996?",
    "difficulty": "hard"
  },
  "b2d0c4b4-ac68-491c-a561-60848ea34fb6": {
    "category": "Entertainment: Music",
    "answers": [
      "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
      "Night Fever",
      "The Chain",
      "Staying Alive"
    ],
    "correctAnswer": "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
    "id": "b2d0c4b4-ac68-491c-a561-60848ea34fb6",
    "question": "Madonna&#039;s song &quot;Hung Up&quot; includes a piece from which popular 70s song?",
    "difficulty": "medium"
  },
  "83cea789-1c6d-4757-a571-c79a14a9133f": {
    "category": "General Knowledge",
    "answers": ["Liver", "Skin", "Heart", "large Intestine"],
    "correctAnswer": "Skin",
    "id": "83cea789-1c6d-4757-a571-c79a14a9133f",
    "question": "What is the largest organ of the human body?",
    "difficulty": "easy"
  },
  "c051d919-2618-4f4d-bb42-5e13bdf58423": {
    "category": "General Knowledge",
    "answers": ["Vanilla", "Saffron", "Cinnamon", "Cardamom"],
    "correctAnswer": "Saffron",
    "id": "c051d919-2618-4f4d-bb42-5e13bdf58423",
    "question": "What is the world&#039;s most expensive spice by weight?",
    "difficulty": "medium"
  },
  "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8": {
    "category": "Entertainment: Video Games",
    "answers": ["Diamond/Pearl", "Ruby/Sapphire", "X/Y", "Black/White"],
    "correctAnswer": "Diamond/Pearl",
    "id": "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8",
    "question": "Which Pokemon generation did the fan-named &quot;Masuda Method&quot; first appear in? ",
    "difficulty": "easy"
  },
  "4db885c9-d4c1-4515-9e20-db71d6557cd6": {
    "category": "Entertainment: Video Games",
    "answers": [
      "Tateyuki Shigaraki",
      "Manfred Von Karma",
      "Gregory Edgeworth",
      "Bansai Ichiyanagi"
    ],
    "correctAnswer": "Bansai Ichiyanagi",
    "id": "4db885c9-d4c1-4515-9e20-db71d6557cd6",
    "question": "In the &quot;Ace Attorney&quot; series, who was the truly responsible of the forging of the autopsy report of the pivotal IS-7 incident?",
    "difficulty": "hard"
  },
}

// mock questions stats
export const mockDifficultyStats: TStat = {
  medium: 3,
  hard: 2,
  easy: 3,
  "all difficulties": 8
}

// mock category stats for different difficulties
export const mockCategoryStats: TStat = {
  "Entertainment: Film": 2,
  "Entertainment: Video Games": 2,
  "General Knowledge": 2,
  "Entertainment: Music": 2,
  "all categories": 8
}

export const mockEasyStats: TStat = {
  "Entertainment: Film": 1,
  "Entertainment: Video Games": 1,
  "General Knowledge": 1,
  "all categories": 3
}

export const mockHardStats: TStat = {
  "Entertainment: Music": 1,
  "Entertainment: Video Games": 1,
  "all categories": 2
}

export const mockMediumStats: TStat = {
  "Entertainment: Film": 1,
  "General Knowledge": 1,
  "Entertainment: Music": 1,
  "all categories": 3
}

// mock question
export const mockQuestion: IQuestion = {
  question: "Mock &lt;Question&gt;",
  id: "mockId",
  difficulty: "easy",
  correctAnswer: "Mock Correct Answer",
  category: "Mock Category",
  answers: [
    "Mocha &amp; Chai",
    "React &#124; Vue &#124; Angular",
    "Mock Aswer 3",
    "Mock Correct Answer",
  ],
};
