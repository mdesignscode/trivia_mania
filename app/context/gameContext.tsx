/* Handle all logic for game page */
"use client";

import { GlobalContext } from "@/app/context/globalContext";
import useFetchQuestionsList from "@/hooks/fetchQuestionsList";
import { IUpdateUserStatsRequest } from "@/models/customRequests";
import {
  CategoryStat,
  DifficultyStat,
  IQuestion,
  IUserStats,
  initialStat,
} from "@/models/interfaces";
import {
  ANSWERED_QUESTIONS,
  PROGRESS,
  UNSAVED_DATA,
  clearQuestionData,
} from "@/utils/localStorage_utils";
import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useRouter } from "next/navigation";
import {
  Dispatch,
  ReactElement,
  SetStateAction,
  createContext,
  useContext,
  useEffect,
  useState,
} from "react";

export interface IGameContext {
  playerStats: IUserStats;
  submitProgress: (testEnv?: boolean) => void;
  updateProgress: (question: IQuestion, answer: string) => void;
  questions: IQuestion[];
  nextQuestionsSet: () => void;
  questionsLength: number;
  questionIndex: number;
  incrementIndex: () => void;
  poolIndex: number;
  startPlaying: boolean;
  setStartPlaying: Dispatch<SetStateAction<boolean>>;
  error: string;
  setQuestionsPool: (state: IQuestion[]) => void;
  answeredQuestions: string[];
  hasSubmit: boolean;
}

export const initialGameContext: IGameContext = {
  playerStats: initialStat,
  submitProgress: () => {},
  updateProgress: () => {},
  questions: [],
  nextQuestionsSet: () => {},
  questionsLength: 0,
  questionIndex: 1,
  incrementIndex: () => {},
  poolIndex: 0,
  setStartPlaying: () => {},
  startPlaying: false,
  error: "",
  setQuestionsPool: () => {},
  answeredQuestions: [],
  hasSubmit: false,
};

export const GameContext = createContext<IGameContext>(initialGameContext);

export function GameProvider({ children }: { children: ReactElement }) {
  const [playerStats, setPlayerStats] = useState<IUserStats>(initialStat);
  const [hasSubmit, setHasSubmit] = useState(false);

  const [answeredQuestions, setAnsweredQuestions] = useState<string[]>([]);

  const [shouldSubmit, setShouldSubmit] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");

  const [startPlaying, setStartPlaying] = useState(false);

  const router = useRouter();

  const { storageIsAvailable, triviaUser } = useContext(GlobalContext);

  const {
    questionIndex,
    poolIndex,
    nextQuestionsSet,
    questions,
    incrementIndex,
    questionsLength,
    setQuestionsPool,
  } = useFetchQuestionsList();

  const { isError, isFetched, error } = useQuery({
    queryKey: ["submitProgress"],
    queryFn: async () => {
      const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
      const url = `${baseUrl}users/updateStats`;

      const { data } = await axios.post(url, {
        stats: playerStats,
        id: triviaUser?.id,
        answeredQuestions: answeredQuestions.filter(Boolean),
      } as IUpdateUserStatsRequest);

      return data;
    },
    enabled: shouldSubmit,
  });

  // restore previous progress state
  useEffect(() => {
    if (storageIsAvailable) {
      // get previous answered questions and progress
      const localAnsweredQuestions =
        localStorage.getItem(ANSWERED_QUESTIONS) || "";
      const localProgress =
        localStorage.getItem(PROGRESS) || JSON.stringify(initialStat);

      // set state
      setPlayerStats(() => JSON.parse(localProgress));
      setAnsweredQuestions(() => localAnsweredQuestions.split(","));
    }
  }, [storageIsAvailable]);

  useEffect(() => {
    if (isError) {
      setErrorMessage((error as any).message);
      return;
    }

    if (isFetched && triviaUser) {
      if (storageIsAvailable) {
        clearQuestionData();
      }
      setHasSubmit(true);
      router.push("/stats" + triviaUser.id);
    }
  }, [error, isError, isFetched, router, storageIsAvailable, triviaUser]);

  const updateProgress = (question: IQuestion, answer: string) => {
    setAnsweredQuestions((state) => {
      const truthyAnswered = state.filter(Boolean);
      const newState = truthyAnswered.length
        ? [...state, question.id]
        : [question.id];

      if (storageIsAvailable) {
        localStorage.setItem(
          ANSWERED_QUESTIONS,
          Array.from(new Set(newState)).join(",")
        );
      }
      return newState;
    });

    setPlayerStats((state) => {
      const newState = {
        ...JSON.parse(JSON.stringify(state)),
      };
      const isCorrect = answer === question.correctAnswer;

      // difficulty
      const difficultyKey = question.difficulty;
      const difficultyStat = state[difficultyKey] as DifficultyStat;
      if (state[difficultyKey]) {
        // increment old state
        newState[difficultyKey].answered = difficultyStat.answered + 1;
        newState[difficultyKey].correctAnswered = isCorrect
          ? difficultyStat.correctAnswered + 1
          : difficultyStat.correctAnswered;
      } else {
        // create new object
        newState[difficultyKey] = {
          answered: 1,
          correctAnswered: isCorrect ? 1 : 0,
        };
      }

      // category
      const categoryKey = question.category;
      const categoryStat = state[categoryKey] as CategoryStat;

      if (categoryStat) {
        const categoryDifficulty = categoryStat[
          difficultyKey
        ] as DifficultyStat;
        if (categoryDifficulty) {
          // increment old category difficulty
          newState[categoryKey][difficultyKey].answered =
            categoryDifficulty.answered + 1;
          newState[categoryKey][difficultyKey].correctAnswered = isCorrect
            ? categoryDifficulty.correctAnswered + 1
            : categoryDifficulty.correctAnswered;
        } else {
          // set new category difficulty
          newState[categoryKey][difficultyKey] = {
            answered: 1,
            correctAnswered: isCorrect ? 1 : 0,
          };
        }
      } else {
        newState[categoryKey] = {
          [difficultyKey]: {
            answered: 1,
            correctAnswered: isCorrect ? 1 : 0,
          },
        };
      }

      // increment total
      newState.total.answered++;
      newState.total.correctAnswered += isCorrect ? 1 : 0;

      if (storageIsAvailable) {
        localStorage.setItem(PROGRESS, JSON.stringify(newState));
        localStorage.setItem(UNSAVED_DATA, JSON.stringify(newState));
      }

      return newState;
    });
  };

  const submitProgress = () => {
    if (triviaUser) {
      setShouldSubmit(true);
    }
  };

  const store: IGameContext = {
    submitProgress,
    updateProgress,
    questionIndex,
    incrementIndex,
    playerStats,
    questions,
    questionsLength,
    nextQuestionsSet,
    poolIndex,
    startPlaying,
    setStartPlaying,
    error: errorMessage,
    setQuestionsPool,
    answeredQuestions,
    hasSubmit,
  };

  return <GameContext.Provider value={store}>{children}</GameContext.Provider>;
}
