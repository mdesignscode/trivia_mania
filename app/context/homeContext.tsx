/* Handles all logic for home page */
"use client";
import { GlobalContext } from "@/context/globalContext";
import useInitialStats from "@/hooks/initialStats";
import { TStatsRequest } from "@/models/customRequests";
import axios from "axios";
import {
  Dispatch,
  SetStateAction,
  createContext,
  useContext,
  useEffect,
  useState,
} from "react";

interface ICurrentUI {
  welcome: boolean;
  difficulties: boolean;
  categories: boolean;
  play: boolean;
}

export interface IHomeContext {
  categoryStats: Record<string, number>;
  fetchingCategories: boolean;
  getDifficultyCategoriesStats: (difficulty: string) => Promise<void>;
  setFetchingCategories: Dispatch<SetStateAction<boolean>>;
  currentUI: ICurrentUI;
  setCurrentUI: Dispatch<SetStateAction<ICurrentUI>>;
  showCategories: boolean
}

export const defaultHomeContext: IHomeContext = {
  categoryStats: {},
  fetchingCategories: true,
  getDifficultyCategoriesStats: () => Promise.resolve(),
  setFetchingCategories: () => {},
  currentUI: {
    welcome: true,
    difficulties: false,
    categories: false,
    play: false,
  },
  setCurrentUI: () => {},
  showCategories: false
};

export const HomeContext = createContext<IHomeContext>(defaultHomeContext);

export function HomeProvider({ children }: { children: React.ReactNode }) {
  // home state
  const { categoryStats: initialCategoryStats, categoriesLoading } =
    useInitialStats();

  const [categoryStats, setCategoryStats] = useState<Record<string, number>>(
    {}
  );

  // display different UI's
  const [currentUI, setCurrentUI] = useState<ICurrentUI>({
    welcome: true,
    difficulties: false,
    categories: false,
    play: false,
  });
  const showCategories = currentUI.categories && !currentUI.welcome;

  const [fetchingCategories, setFetchingCategories] = useState(true);
  const {
    userStatus: { user },
  } = useContext(GlobalContext);

  useEffect(() => {
    setFetchingCategories(categoriesLoading);
    setCategoryStats(initialCategoryStats);
  }, [categoriesLoading, initialCategoryStats]);

  // fetch category stats based on difficulty
  async function getDifficultyCategoriesStats(difficulty: string) {
    try {
      // load categories
      setFetchingCategories(true);

      const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
      const url = baseUrl + "questions/stats";

      const { data } = await axios.post(url, {
        recordType: "categories",
        difficulty,
        userId: user?.id,
      } as TStatsRequest);
      setCategoryStats(data);

      // display categories
      setFetchingCategories(false);
    } catch (error) {
      console.log(error);
    }
  }

  // store object
  const store: IHomeContext = {
    showCategories,
    categoryStats,
    fetchingCategories,
    getDifficultyCategoriesStats,
    setFetchingCategories,
    setCurrentUI,
    currentUI,
  };

  return <HomeContext.Provider value={store}>{children}</HomeContext.Provider>;
}
