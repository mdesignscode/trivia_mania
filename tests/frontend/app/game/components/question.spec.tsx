import { GameProvider } from "@/app/context/gameContext";
import { GlobalProvider } from "@/app/context/globalContext";
import RenderQuestion, {
  IRenderQuestion,
  decodeHTMLEntities,
} from "@/app/game/components/question";
import useFetchQuestionsList from "@/hooks/fetchQuestionsList";
import { QUESTION_ANSWERED } from "@/utils/localStorage_utils";
import { mockQuestion } from "@/utils/mockData";
import {
  mockSubmitProgress,
  renderGameContext,
} from "@/utils/test_game_context";
import { screen, userEvent, render, renderHook } from "@/utils/test_utils";

// spy functions
const mockHandleNextQuestion = jest.fn(),
  mockHandleTimesUp = jest.fn(),
  mockHandleViewProgress = jest.fn(),
  mockHandleUserAnswer = jest.fn(),
  mockContinueLater = jest.fn();

// mock props
const mockProps: IRenderQuestion = {
  userAnswer: "",
  timesUp: false,
  timerHasStarted: true,
  index: 1,
  CTA: "Next Question",
  answerFeedback: [false, false, false, false],
  handleNextQuestion: mockHandleNextQuestion,
  handleTimesUp: mockHandleTimesUp,
  handleUserAnswer: mockHandleUserAnswer,
  handleViewProgress: mockHandleViewProgress,
  questionObj: mockQuestion,
  handleContinueLater: mockContinueLater,
};

describe("RenderQuestion component", () => {
  it("Should render a question for a user to answer", async () => {
    // render question
    renderGameContext(<RenderQuestion {...mockProps} />);

    // get elements from DOM
    const container = await screen.findByTestId("question-container"),
      question = await screen.findByText(
        decodeHTMLEntities(mockQuestion.question)
      ),
      easy = await screen.findByText("easy"),
      category = await screen.findByText(mockQuestion.category),
      timer = await screen.findByTestId("timer-container");

    // assert elements are in DOM
    expect(container).toBeInTheDocument();
    expect(question).toBeInTheDocument();
    expect(easy).toBeInTheDocument();
    expect(category).toBeInTheDocument();
    expect(timer).toBeInTheDocument();

    // assert answer buttons are in DOM
    for (const entity of mockQuestion.answers) {
      const answer = decodeHTMLEntities(entity);
      const button = await screen.findByText(answer);

      expect(button).toBeInTheDocument();
    }
  });

  it("Should gracefully handle user answer", async () => {
    // setup user interaction
    const user = userEvent.setup();

    // render question
    renderGameContext(<RenderQuestion {...mockProps} />);

    // find an answer button to click
    const randomValue = Math.floor(Math.random() * 4),
      randomAnswerValue = decodeHTMLEntities(mockQuestion.answers[randomValue]);
    const randomAnswer = await screen.findByText(randomAnswerValue);

    expect(randomAnswer).toBeInTheDocument();

    // answer question
    await user.click(randomAnswer);

    // assert answer handling
    expect(mockHandleUserAnswer).toBeCalledWith(randomAnswerValue, randomValue);
  });

  it("Should save a user's progress on clicking `View Progress` or `Submit Results`", async () => {
    // setup user interaction
    const user = userEvent.setup();

    // render question with props to simulate `View Progress`
    const { rerender } = renderGameContext(
      <RenderQuestion
        {...{
          ...mockProps,
          CTA: "Continue Playing",
          timesUp: true,
          timerHasStarted: false,
        }}
      />
    );

    // find `View Progress` button
    const viewProgressButton = await screen.findByText("View Progress");
    expect(viewProgressButton).toBeInTheDocument();

    // click button
    await user.click(viewProgressButton);

    expect(mockHandleViewProgress).toBeCalled();

    // rerender question with props to simulate `Submit Results`
    rerender(
      <RenderQuestion
        {...{
          ...mockProps,
          CTA: "Submit Results",
          timesUp: true,
          timerHasStarted: false,
        }}
      />
    );

    // find `Submit Results` button
    const sumbitResultsButton = await screen.findByText("Submit Results");
    expect(sumbitResultsButton).toBeInTheDocument();

    // click button
    await user.click(sumbitResultsButton);

    expect(mockHandleNextQuestion).toBeCalled();
  });

  it("Shifts the questions pool to the right and saves a user's progress on clicking `Continue Later`", async () => {
    // setup user interaction
    const user = userEvent.setup();

    // renderGameContext(<RenderQuestion {...mockProps} />);

    render(
      <GlobalProvider>
        <GameProvider>
          <RenderQuestion {...mockProps} />
        </GameProvider>
      </GlobalProvider>
    );

    const continueLaterButton = await screen.findByTestId(
      "continue-later-button"
    );
    expect(continueLaterButton).toBeInTheDocument();

    await user.click(continueLaterButton);

    expect(mockContinueLater).toBeCalled();
  });

  it("snapshot matches", () => {
    // render component
    const { baseElement } = renderGameContext(
      <RenderQuestion {...mockProps} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
