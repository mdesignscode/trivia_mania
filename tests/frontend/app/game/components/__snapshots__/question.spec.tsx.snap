// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderQuestion component snapshot matches 1`] = `
<body>
  <div>
    <div>
      <div
        class="sc-grXZZQ lakGlY question col gap-7 rounded-lg p-6"
        data-testid="question-container"
      >
        <div
          class="flex justify-between items-center"
        >
          <h1
            class="text-xl"
          >
            Mock Category
          </h1>
          <p
            class="text-xl"
            style="color: green;"
          >
            easy
          </p>
        </div>
        <h1
          class="text-2xl"
        >
          Mock &lt;Question&gt;
        </h1>
        <svg
          data-testid="timer-container"
          style="border: 2px solid green; border-radius: 5px;"
          viewBox="0 0 100 5"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            class="transition-all"
            stroke="green"
            stroke-width="20"
            x1="0"
            x2="0"
            y1="5"
            y2="5"
          />
          30
        </svg>
        <div
          class="question_options grid grid-cols-2 grid-rows-2 gap-4"
        >
          <span
            data-testid="Mocha & Chai"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mocha & Chai"
            >
              <span
                data-testid="feedback-incorrect-0"
              />
              <p>
                Mocha & Chai
              </p>
            </button>
          </span>
          <span
            data-testid="React | Vue | Angular"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="React | Vue | Angular"
            >
              <span
                data-testid="feedback-incorrect-1"
              />
              <p>
                React | Vue | Angular
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Aswer 3"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mock Aswer 3"
            >
              <span
                data-testid="feedback-incorrect-2"
              />
              <p>
                Mock Aswer 3
              </p>
            </button>
          </span>
          <span
            data-testid="Mock Correct Answer"
          >
            <button
              class="sc-bcPKhP fVQoNj flex w-full justify-center gap-2 items-center animate__animated"
              id="Mock Correct Answer"
            >
              <span
                data-testid="feedback-correct-3"
              />
              <p>
                Mock Correct Answer
              </p>
            </button>
          </span>
        </div>
        <span
          data-testid="continue-later-button"
        >
          <button
            class="sc-bcPKhP fVQoNj w-full"
          >
            Continue Later
          </button>
        </span>
        <audio
          id="success"
        >
          <source
            src="/success.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
        <audio
          id="error"
        >
          <source
            src="/error.mp3"
            type="audio/mpeg"
          />
          Your browser does not support the audio element.
        </audio>
      </div>
    </div>
  </div>
</body>
`;
